### PR TITLE
Closes #3086: Add `choice` to random number generators

### LINKED ISSUE #3086: Add `choice` to random number generators
Add `choice` to our generators

### PR DESCRIPTION
**Pull request by stress-tess**
This PR (closes #3086) adds `choice` to random number generators.

I had to make a very slight modification to `sampleDomWeighted` (just moving the increment to be after indexing) to avoid an out of bounds error.

### CODE REVIEW

**Comment by jeremiah-corrado** [comment]
Yep, unfortunately it looks like there is a subtle bug in choice where it can call the module-level `shuffle` proc (which generates a new random seed), rather than the random stream's `shuffle` method. This only comes up in the unweighted case when `replace=false` and the number of samples is greater than `log2(pop_size)`.

<details>
<summary>diff for 1.33...</summary>

```diff
diff --git a/modules/standard/Random.chpl b/modules/standard/Random.chpl
index 51e8d8723f..b5c50ed4c2 100644
--- a/modules/standard/Random.chpl
+++ b/modules/standard/Random.chpl
@@ -948,7 +948,7 @@ module Random {
           }
         } else {
           var indices: [X] int = X;
-          shuffle(indices);
+          stream.shuffle(indices);
           for i in samples.domain {
             samples[i] = (indices[X.dim(0).orderToIndex(i)]);
           }
```

</details>

I'm not sure whether this is worth fixing, but I suppose the solution would be to add a correct implementation of choice to Arkouda.

**Comment by stress-tess** [comment]
> Yep, unfortunately it looks like there is a subtle bug in choice where it can call the module-level `shuffle` proc (which generates a new random seed), rather than the random stream's `shuffle` method. This only comes up in the unweighted case when `replace=false` and the number of samples is greater than `log2(pop_size)`.
> 
> patch for Chapel 1.33...
> I'm not sure whether this is worth fixing, but I suppose the solution would be to add a correct implementation of choice to Arkouda.

thanks so much for running this down for me @jeremiah-corrado!! Unfortunately I do think that most of our users haven't made the jump to `2.0` yet, so I do think adding a correct implementation of choice is prob the way to go even tho that does not seem like the most fun task lol

I was thinking I might skip that test for now and then add the correct choice impl in a separate PR. @Bears-R-Us/arkouda-core-dev , what are y'all's thoughts on that

**Review by ajpotts — APPROVED** [review]
Looks good to me!  I like how you used the chi-squared test :)

**Review by jaketrookman — APPROVED** [review]

File: PROTO_tests/tests/random_test.py [hunk 1]
```diff
@@ -126,6 +127,85 @@ def test_uniform(self):
...
+        # I tested this many times without a set seed, but with no seed
+        # it's expected to fail one out of every ~20 runs given a pval limit of 0.05.
+        rng = ak.random.default_rng(43)
+        num_samples = 10**4
+
+        weights = ak.array([0.25, 0.15, 0.20, 0.10, 0.30])
+        weighted_sample = rng.choice(ak.arange(5), size=num_samples, p=weights)
+
+        # count how many of each category we saw
+        uk, f_obs = ak.GroupBy(weighted_sample).count()
```
[the remark this answers is not in the document]
Comment (stress-tess) [reply PROTO_tests/tests/random_test.py]: updated!

File: PROTO_tests/tests/random_test.py [hunk 2]
```diff
@@ -126,6 +127,85 @@ def test_uniform(self):
...
+        domain_choice = rng.choice(20, 20, replace=False)
+        # since our populations and sample size is the same without replacement,
+        # we should see all values
+        assert (ak.sort(domain_choice) == ak.arange(20)).all()
+
+        # test arrays and not selecting all
+        perm = rng.permutation(100)
+        array_choice = rng.choice(perm, 95, replace=False)
+        # verify all unique
+        _, count = ak.GroupBy(array_choice).count()
```
[the remark this answers is not in the document]
Comment (stress-tess) [reply PROTO_tests/tests/random_test.py]: updated!

**Pull request merged**